pkg/controller/master/virtualmachine/vm_controller.go

```diff
@@ -377,29 +378,74 @@ func (h *VMController) SetHaltIfInsufficientResourceQuota(_ string, vm *kubevirt
 		return vm, nil
 	}
 
-	if err := h.vmrCalculator.CheckIfVMCanStartByResourceQuota(vm); err == nil {
-		return vm, nil
+	err := h.vmrCalculator.CheckIfVMCanStartByResourceQuota(vm)
+	if err == nil {
+		if vm.Status.PrintableStatus == kubevirtv1.VirtualMachineStatusStarting {
+			logrus.Debugf("VM %s in namespace %s is starting", vm.Name, vm.Namespace)
+
+			// Since VMI would wait for Pod to run status,
+			// cause the VM is still in Starting,
+			// that recalculates the VM resource quota each time until the VM is non-Starting.
+			h.vmClient.EnqueueAfter(vm.Namespace, vm.Name, 5*time.Second)
+			return vm, nil
+		}
+		return vm, h.cleanUpInsufficientResourceAnnotation(vm)
 	} else if !rqutils.IsInsufficientResourceError(err) {
 		return nil, err
 	}
 
-	if err := h.stopVM(vm); err != nil {
-		return nil, err
+	return vm, h.stopVM(vm, err.Error())
+}
+
+func (h *VMController) stopVM(vm *kubevirtv1.VirtualMachine, errMsg string) error {
+	return stopVM(h.vmClient, h.recorder, vm, errMsg)
+}
+
+func (h *VMController) cleanUpInsufficientResourceAnnotation(vm *kubevirtv1.VirtualMachine) error {
+	if vm.Annotations == nil {
+		return nil
+	} else if _, ok := vm.Annotations[util.AnnotationInsufficientResourceQuota]; !ok {
+		return nil
 	}
 
-	h.recorder.Eventf(vm, corev1.EventTypeWarning, "InsufficientResourceQuota", "Set runStrategy to %s", kubevirtv1.RunStrategyHalted)
-	return vm, nil
+	if runStrategy, err := vm.RunStrategy(); err != nil {
+		return err
+	} else if runStrategy == kubevirtv1.RunStrategyHalted {
+		logrus.Debugf("skip cleaning up insufficient resource annotation, VM %s in namespace %s is halted.", vm.Name, vm.Namespace)
+		return nil
+	}
+
+	if vmi, err := h.vmiCache.Get(vm.Namespace, vm.Name); err != nil && !apierrors.IsNotFound(err) {
+		return err
+	} else if err != nil || !vmi.IsRunning() {
+		logrus.Debugf("skip cleaning up insufficient resource annotation, VM %s in namespace %s is not running.", vm.Name, vm.Namespace)
+		return nil
+	}
```

Comment:
Prefer to reduce the use of `else if`, which will makes the code difficult to understand. Unless the condition of `if` is on the same variable as the condition of `else if`
```go
	vmi, err := h.vmiCache.Get(vm.Namespace, vm.Name)
	if err != nil && !apierrors.IsNotFound(err) {
		return err
	}
	if vmi == nil || !vmi.IsRunning() {
		logrus.Debugf("skip cleaning up insufficient resource annotation, VM %s in namespace %s is not running.", vm.Name, vm.Namespace)
		return nil
	}
```